src/internal_tricks.rs

```diff
@@ -35,15 +35,45 @@ macro_rules! pyo3_exception {
 #[derive(Debug)]
 pub(crate) struct NulByteInString(pub(crate) &'static str);
 
+#[derive(Copy, Clone)]
+pub(crate) enum MaybeLeaked {
```

Comment:
Isn't this actually `MaybeOwned::Owned(*mut c_char)` because the point of the type is to not leak things?

Also could this also be something like

```rust
pub(crate) enum MaybeOwned {
  Static(&'static CStr),
  Owned(CString),
}
```

or does this violate any aliasing guarantees?

Reply:
I went with my approach over this for some reasons:
- the type is smaller (doesn't really matter, I suppose)
- I don't think aliasing `CString`  is a problem in practice but using it would make `ClosureDestructor` (more of) a self referential struct, so I'd really prefer to use raw pointers 
- it's a lot easier to create a `*const c_char` at compile time than a `&'static CStr` (msrv 😭 )

Reply:
Another option that comes to mind is to not store the `Static` variant at all (we do not need it for destruction) and use a signature similar to `as_method_def` above for `extract_cstr_or_leak_cstring`, e.g. make it run

```rust
Result<(*const c_char, ManuallyDrop<Option<CString>>), NulByteInString>
```

and use `Option<CString>` inside of `PyMethodDefDestructor` (which is a sort of "double-option" ATM).